File: bot.php
<?php
declare(strict_types=1);

require 'madeline.php';

const DOWNLOADS_DIR = __DIR__ . '/downloads';
const DB_PATH       = __DIR__ . '/files.db';

$token = getenv('BOT_TOKEN');
if (!$token) die("BOT_TOKEN env yo'q\n");

/* ─── DB ──────────────────────────────────────────────────── */
$db = new PDO('sqlite:' . DB_PATH, options: [
    PDO::ATTR_ERRMODE            => PDO::ERRMODE_EXCEPTION,
    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
]);
$db->exec("PRAGMA journal_mode=WAL");
$db->exec("PRAGMA synchronous=NORMAL");
$db->exec("
    CREATE TABLE IF NOT EXISTS files (
        id         INTEGER PRIMARY KEY AUTOINCREMENT,
        message_id INTEGER NOT NULL,
        sender_id  INTEGER,
        file_name  TEXT    NOT NULL,
        duration   INTEGER NOT NULL DEFAULT 0,
        size       INTEGER NOT NULL DEFAULT 0,
        path       TEXT    NOT NULL,
        created_at TEXT    NOT NULL,
        UNIQUE(message_id)
    )
");

/* ─── HELPERS ─────────────────────────────────────────────── */
function extractMeta(array $media): array
{
    $fileName = 'audio_' . time() . '.ogg';
    $duration = 0;
    $size     = (int) ($media['document']['size'] ?? 0);

    foreach ($media['document']['attributes'] ?? [] as $attr) {
        if ($attr['_'] === 'documentAttributeFilename') {
            $fileName = $attr['file_name'];
        }
        if ($attr['_'] === 'documentAttributeAudio') {
            $duration = (int) ($attr['duration'] ?? 0);
        }
    }
    return compact('fileName', 'duration', 'size');
}

function uniquePath(string $dir, string $fileName): string
{
    $base = pathinfo($fileName, PATHINFO_FILENAME);
    $ext  = pathinfo($fileName, PATHINFO_EXTENSION);
    $path = $dir . '/' . $fileName;
    $i    = 1;
    while (file_exists($path)) {
        $path = $dir . '/' . $base . '_' . $i . ($ext ? ".$ext" : '');
        $i++;
    }
    return $path;
}

/* ─── BOT ─────────────────────────────────────────────────── */
if (!is_dir(DOWNLOADS_DIR)) mkdir(DOWNLOADS_DIR, 0755, true);

$MadelineProto = new \danog\MadelineProto\API('bot.session');
$MadelineProto->botLogin($token);

echo "[OK] Bot started\n";

$offset = 0;

while (true) {
    try {
        $updates = $MadelineProto->getUpdates(['offset' => $offset, 'limit' => 100]);

        foreach ($updates as $update) {
            $offset = $update['update_id'] + 1;

            $msg = $update['update']['message'] ?? null;
            if (!$msg) continue;

            $messageId = (int) ($msg['id'] ?? 0);
            $senderId  = (int) ($msg['from_id']['user_id'] ?? 0);
            $media     = $msg['media'] ?? null;

            if (!$media || !in_array($media['_'], [
                'messageMediaDocument',
                'messageMediaAudio',
                'messageMediaVoice',
            ], true)) {
                $MadelineProto->messages->sendMessage([
                    'peer'            => $msg['peer_id'],
                    'message'         => '🎵 Menga faqat audio yuboring.',
                    'reply_to_msg_id' => $messageId,
                ]);
                continue;
            }

            // Duplicate check
            $stmt = $db->prepare("SELECT 1 FROM files WHERE message_id=? LIMIT 1");
            $stmt->execute([$messageId]);
            if ($stmt->fetchColumn()) {
                echo "[SKIP] msg=$messageId\n";
                continue;
            }

            ['fileName' => $fileName, 'duration' => $duration, 'size' => $size]
                = extractMeta($media);

            $savePath = uniquePath(DOWNLOADS_DIR, $fileName);

            $MadelineProto->messages->sendMessage([
                'peer'            => $msg['peer_id'],
                'message'         => '⏳ Yuklanmoqda...',
                'reply_to_msg_id' => $messageId,
            ]);

            $MadelineProto->downloadToFile($media, $savePath);

            $db->prepare("
                INSERT OR IGNORE INTO files
                    (message_id, sender_id, file_name, duration, size, path, created_at)
                VALUES (?, ?, ?, ?, ?, ?, ?)
            ")->execute([$messageId, $senderId, $fileName, $duration, $size, $savePath, date('Y-m-d H:i:s')]);

            echo "[OK] $fileName\n";

            $MadelineProto->messages->sendMessage([
                'peer'            => $msg['peer_id'],
                'message'         => "✅ Saqlandi!\n\n📁 $fileName\n⏱ {$duration}s\n📦 " . round($size/1024/1024, 2) . " MB",
                'reply_to_msg_id' => $messageId,
            ]);
        }

        sleep(2);

    } catch (\danog\MadelineProto\Exception $e) {
        echo "[MTProto] " . $e->getMessage() . "\n";
        sleep(5);
    } catch (PDOException $e) {
        echo "[DB] " . $e->getMessage() . "\n";
        sleep(5);
    } catch (Throwable $e) {
        echo "[ERROR] " . $e->getMessage() . "\n";
        sleep(5);
    }
}

File: bot.py
import os
import sqlite3
import asyncio
from datetime import datetime
from pyrogram import Client, filters
from pyrogram.types import Message

BOT_TOKEN = os.getenv("8680623688:AAHaXKiRee6hY2ZF8GhXyH156A9iexIn6BY")
API_ID    = int(os.getenv("27605865"))
API_HASH  = os.getenv("f76cf301b264391c6ed01747638adecc")

DOWNLOADS_DIR = "downloads"
DB_PATH       = "files.db"

os.makedirs(DOWNLOADS_DIR, exist_ok=True)

# ── DB ──────────────────────────────────────────────────────
def get_db():
    db = sqlite3.connect(DB_PATH)
    db.execute("PRAGMA journal_mode=WAL")
    db.execute("""
        CREATE TABLE IF NOT EXISTS files (
            id         INTEGER PRIMARY KEY AUTOINCREMENT,
            message_id INTEGER NOT NULL,
            sender_id  INTEGER,
            file_name  TEXT    NOT NULL,
            duration   INTEGER NOT NULL DEFAULT 0,
            size       INTEGER NOT NULL DEFAULT 0,
            path       TEXT    NOT NULL,
            created_at TEXT    NOT NULL,
            UNIQUE(message_id)
        )
    """)
    db.commit()
    return db

def is_duplicate(db, message_id: int) -> bool:
    cur = db.execute("SELECT 1 FROM files WHERE message_id=? LIMIT 1", (message_id,))
    return cur.fetchone() is not None

def save_record(db, data: dict):
    db.execute("""
        INSERT OR IGNORE INTO files
            (message_id, sender_id, file_name, duration, size, path, created_at)
        VALUES
            (:message_id, :sender_id, :file_name, :duration, :size, :path, :created_at)
    """, data)
    db.commit()

def unique_path(file_name: str) -> str:
    base, ext = os.path.splitext(file_name)
    path = os.path.join(DOWNLOADS_DIR, file_name)
    i = 1
    while os.path.exists(path):
        path = os.path.join(DOWNLOADS_DIR, f"{base}_{i}{ext}")
        i += 1
    return path

# ── BOT ─────────────────────────────────────────────────────
app = Client("bot", bot_token=BOT_TOKEN, api_id=API_ID, api_hash=API_HASH)
db  = get_db()

@app.on_message(filters.audio | filters.voice | filters.document)
async def handle_audio(client: Client, msg: Message):
    # Document bo'lsa faqat audio mime type
    if msg.document:
        mime = msg.document.mime_type or ""
        if not mime.startswith("audio/"):
            await msg.reply("🎵 Menga faqat audio, voice yoki audio fayl yuboring.")
            return

    if is_duplicate(db, msg.id):
        print(f"[SKIP] msg={msg.id}")
        return

    media = msg.audio or msg.voice or msg.document
    file_name = getattr(media, "file_name", None) or f"audio_{msg.id}.ogg"
    duration  = getattr(media, "duration", 0) or 0
    size      = getattr(media, "file_size", 0) or 0

    await msg.reply("⏳ Yuklanmoqda...")

    save_path = unique_path(file_name)

    try:
        await client.download_media(msg, file_name=save_path)

        save_record(db, {
            "message_id": msg.id,
            "sender_id":  msg.from_user.id if msg.from_user else 0,
            "file_name":  file_name,
            "duration":   duration,
            "size":       size,
            "path":       save_path,
            "created_at": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        })

        print(f"[OK] {file_name}")

        await msg.reply(
            f"✅ Saqlandi!\n\n"
            f"📁 {file_name}\n"
            f"⏱ {duration}s\n"
            f"📦 {round(size / 1024 / 1024, 2)} MB"
        )

    except Exception as e:
        print(f"[ERROR] {e}")
        await msg.reply("❌ Xatolik yuz berdi.")

@app.on_message(filters.text)
async def handle_text(client: Client, msg: Message):
    await msg.reply("🎵 Menga faqat audio, voice yoki audio fayl yuboring.")

# ── RUN ─────────────────────────────────────────────────────
if __name__ == "__main__":
    print("[OK] Bot started")
    app.run()